projects/cfengine/string_fuzzer.c

```diff
@@ -0,0 +1,28 @@
+/* Copyright 2021 Google LLC
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <stdint.h>
+#include <string.h>
+#include <stdlib.h>
+#include <string_expressions.h>
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size){
+        char *new_str = (char *)malloc(size+1);
+        if (new_str == NULL){
+                return 0;
+        }
+        memcpy(new_str, data, size);
+        new_str[size] = '\0';
```

Comment:
Can `data` have zero-bytes? Is it possible to instruct the fuzzer to not insert zeroes, as that might be an assumption at this point?

Reply:
Yes, LibFuzzer (the fuzzing engine used for this harness) can generate zero-bytes.

It is also possible to instruct the harness to not add those to the input.

Why would we need to do that?

Reply:
We are testing an internal function / an internal part of the parser. It might be the case that this string has already been checked for zero-bytes, and we don't want this specific code to re-check / handle that case. I'm not too familiar with this part of the codebase, but I would guess that this is the case here, that the parser will never send a string with a zero to this function. (But there is no assert there, so I don't really know this).

I want the fuzzer to find issues which can happen in the real world, i.e. either test the full parser, or test internal parts, but with some restrictions depending on the context.

So, I would suggest to start with instructing libfuzz to not generate zero-bytes in this case. In the end, I think it will lead to finding more important problems.

If we test the full parser, we can definitely have the fuzzer insert zeroes as well.

Reply:
Thank you for elaborating. I will just keep this draft open and check during the week if we can hit the full parser in this iteration before merging. 
Before I do that, could you confirm that you are interested in fuzzing CFEngine continuously?

Reply:
@AdamKorcz Yes we are interested.

Starting with a small target like this makes a lot of sense, I don't think you have to or should set up fuzzing of the full parser, but I think for an internal function like this, we should add the limitation of no zero-bytes, as explained above.

Reply:
That sounds good, thank you for the feedback. I will amend the fuzzer to prevent zero-bytes being passed to the target.

Reply:
@olehermanse I updated the fuzzer. Was that what you had in mind?

Reply:
I don't know what the return code (`1`) means here, but I guess that will work. Have you tested it?

Reply:
It worked fine locally but I changed the fuzzer to return `0` in all conditions.

Reply:
@olehermanse Kind reminder.

Reply:
@AdamKorcz You didn't answer the other comment, I asked whether we will get bug reports if it hits an assert (and it fails).

Reply:
Apologies. 

> I asked whether we will get bug reports if it hits an assert (and it fails)

In this case you will get bug reports.

Reply:
Thanks for the clarification.